crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
+    dynamic_paths: Option<ordermap::set::Iter<'db, SearchPathRoot>>,
+}
+
+impl<'db> SearchPathIterator<'db> {
+    fn new(db: &'db dyn Db) -> Self {
+        let static_paths = module_resolver_settings(db)
+            .search_path_settings
+            .static_search_paths()
+            .into_iter();
+
+        Self {
+            db,
+            static_paths,
+            dynamic_paths: None,
+        }
+    }
+}
+
+impl<'db> Iterator for SearchPathIterator<'db> {
+    type Item = &'db SearchPathRoot;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let SearchPathIterator {
+            db,
+            static_paths,
+            dynamic_paths,
+        } = self;
+
+        static_paths.next().or_else(|| {
+            dynamic_paths
+                .get_or_insert_with(|| dynamic_module_resolution_paths(*db).into_iter())
+                .next()
+        })
+    }
+}
 
-impl Deref for OrderedSearchPaths {
-    type Target = [Arc<ModuleResolutionPathBuf>];
+impl<'db> FusedIterator for SearchPathIterator<'db> {}
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+fn iter_module_search_paths(db: &dyn Db) -> SearchPathIterator {
+    SearchPathIterator::new(db)
+}
+
+/// Represents a single `.pth` file in a `site-packages` directory.
+/// One or more lines in a `.pth` file may be a (relative or absolute)
+/// path that represents an editable installation of a package.
+struct PthFile<'db> {
+    db: &'db dyn Db,
+    path: SystemPathBuf,
+    contents: SourceText,
+    site_packages: &'db SystemPath,
+}
+
+impl<'db> PthFile<'db> {
+    fn new(
+        db: &'db dyn Db,
+        pth_path: SystemPathBuf,
+        file: File,
+        site_packages: &'db SystemPath,
+    ) -> Self {
+        Self {
+            db,
+            path: pth_path,
+            contents: source_text(db.upcast(), file),
+            site_packages,
+        }
+    }
+
+    /// Yield paths in this `.pth` file that appear to represent editable installations,
+    /// and should therefore be added as module-resolution search paths.
+    fn iter_editable_installations<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = ModuleResolutionPathBuf> + 'db
+    where
+        'a: 'db,
+    {
+        // Empty lines or lines starting with '#' are ignored by the Python interpreter.
+        // Lines that start with "import " or "import\t" do not represent editable installs at all;
+        // instead, these are files that are executed by Python at startup.
+        // https://docs.python.org/3/library/site.html#module-site
+        self.contents.lines().filter_map(|line| {
+            let line = line.trim();
+            if line.is_empty()
+                || line.starts_with('#')
+                || line.starts_with("import ")
+                || line.starts_with("import\t")
+            {
+                return None;
+            }
+            let possible_editable_install = SystemPath::absolute(
```

Comment:
Isn't the `site_packages` path guaranteed to be absolute? I would prefer guaranteeing that the `site_packages` path is absolute instead of relying on `system.current_directory` here.

Reply:
The `site-packages` path is guaranteed to be absolute here, but without the call to `SystemPath::absolute()` here I think we could end up storing technically-valid-but-somewhat-strange paths as module resolution roots, e.g. `/site-packages/./../foo/bar/baz/src/`. I'd prefer to have that normalized as `/foo/bar/baz/src` before storing it as a module-resolution root (just for debuggability more than anything else).

Reply:
This is because the paths we're trying to discover in `pth` files could be absolute paths, or they could be relative to `site-packages`

Reply:
Makes sense. Can we then pass `site_packages` as the current working directory instead of `system.current_directory`?

Reply:
Ah, yeah that works!